wrc/codegen/merger.py

```diff
@@ -0,0 +1,177 @@
+from wrc.sema.ast import Article, Section, TableOfContent, Guideline, split_rule_number
+from copy import deepcopy
+
+NOTES_INDEX = 0
+TOC_INDEX = 1
+FIRST_ARTICLE_INDEX = 2
+
+
+class BadFormatError(Exception):
+    def __init__(self, message: str):
+        self.message = message + '. Please report this issue to the developer.'
+        super().__init__(message)
+
+
+def compare_article_numbers(num1: str, num2: str) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    This is necessary because:
+    * We need to compare article numbers like '1' and 'A'.
+    * '2' > '10' == True, which is a problem.
+    """
+    is_digit_1 = num1.isdigit()
+    is_digit_2 = num2.isdigit()
+    if is_digit_1 and is_digit_2:
+        res = int(num1) < int(num2)
+    elif is_digit_1 and not is_digit_2:
+        # Only num2 is a letter.
+        res = True
+    elif not is_digit_1 and not is_digit_2:
+        res = num1 < num2
+    else:
+        res = False
+
+    return res
+
+
+def compare_reg_numbers(num1: list, num2: list) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    We need this for the same reasons we need compare_article_numbers().
+    """
+    for n1, n2 in zip(num1, num2):
+        if n1 < n2:
+            return True
+        elif n1 > n2:
+            return False
+
+    return len(num1) < len(num2)
+
+
+def get_reg_number(reg):
+    if isinstance(reg, Guideline):
+        reg = reg.number.split('+')[0]
+        num = split_rule_number(reg)
+    else:
+        num = split_rule_number(reg.number)
```

Comment:
How do you know `reg.number` even exists? At this point in the if-else cascade, all that you know is it is NOT a `Guideline` object. But what else could it be?

Reply:
![Captura de pantalla 2024-03-18 a la(s) 18 52 52](https://github.com/thewca/wca-regulations-compiler/assets/59543882/982fef2b-bf10-4cb5-a552-46f7697a34b3)


This is the structure of the AST. As you can see, some sections are from the `Article` class, and those sections only contain regulations and guidelines.

If you start iterating over the `sections` list of the AST, you will find the first `Article` section, and starting from there, all sections are Articles.

Lines 74 to 76 are meant to check that `FIRST_ARTICLE_INDEX` is indeed the index of the first article on the `sections` list. Then, we iterate over the `Article` sections and we add the content of each of those sections to the stack (see lines 89 and 91).

So the answer is that reg.number should always exist unless something extremely weird is going on with the AST structure.

I can still add a check and throw an error if you think it's necessary. :)

Reply:
What I was trying to say is that it _should_ be impossible for the parameter to be anything other than a Regulation or a Guideline instance because of the previous checks. But I get you. This is how it looks now:

```python3
  if isinstance(reg, Guideline):
      reg = reg.number.split('+')[0]
      num = split_rule_number(reg)
  elif isinstance(reg, Regulation):
      num = split_rule_number(reg.number)
  else:
      raise BadFormatError(f'get_reg_number expected Regulation or Guideline, got {type(reg).__name__}')
```